changefeedccl: Checkpoint changefeeds during backfill.

miretskiy opened the pull request:
Checkpoint changefeed backfill progress.  This is accomplished
by resolved span events for each sub-span of the table.
Those events are aggregated by change frontier processor which checkpoints
spans whose timestamp is above the current high water mark.

Informs #64293

Release Notes: Changefeed can resume during backfill without
loosing too much progress.

Commit:
changefeedccl: Emit resolved events during initial backfill.

`AddResolved` events during initial backfill.

Release Notes: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/66013)
<!-- Reviewable:end -->

stevendanna reviewed (approved):
I left a few nitpicks but overall the code looks good to me.

One question that might be nice to answer in the commit message is why we are putting this checkpoint field in the ChangeAggregatorSpec rather than the progress.

> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66013#-:-MbG40vR6Wtk3Ra6m3Jr:bxwokjg)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @miretskiy)

___
*[pkg/ccl/changefeedccl/kvfeed/scanner.go, line 151 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66013#-MbG1lct-FwEdXTbXu5d:-MbG1lcu9gif6eDDH-sQ:b-tje7o0) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2f8b03c9328d6e134ddd1478c6c512d7699c40c/pkg/ccl/changefeedccl/kvfeed/scanner.go#L151)):*
> ```Go
> 				return err
> 			}
> 			remaining = res.ResumeSpan
> ```

[nit] since we do this below, I think we can drop this line
___
*[pkg/sql/execinfrapb/processors_changefeeds.proto, line 37 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66013#-MbG0Hi2-JOHiUAiIbK6:-MbG0Hi2-JOHiUAiIbK7:b-nznwp5) ([raw file](https://github.com/cockroachdb/cockroach/blob/bbc300dc320aa45099b7f32784d8310cb7b39b29/pkg/sql/execinfrapb/processors_changefeeds.proto#L37)):*
> ```protobuf
>   // WatchCheckpoint is the list of watches that were previously completed to specified
>   // timestamp.
>   repeated Watch watch_checkpoint = 5 [(gogoproto.nullable) = false];
> ```

[nit] for other `repeated` fields we use plural field names.


<!-- Sent from Reviewable.io -->

miretskiy reviewed (commented):
> ChangeAggregatorSpec
Updated.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66013#-:-MbGj_aF6Rfex1bOb0Xf:bk53syq)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @stevendanna)

___
*[pkg/ccl/changefeedccl/kvfeed/scanner.go, line 151 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66013#-MbG1lct-FwEdXTbXu5d:-MbGjlta0yzIYpO5HR-R:b-7d8a1w) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2f8b03c9328d6e134ddd1478c6c512d7699c40c/pkg/ccl/changefeedccl/kvfeed/scanner.go#L151)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

[nit] since we do this below, I think we can drop this line
</blockquote></details>

Yup -- good point
___
*[pkg/sql/execinfrapb/processors_changefeeds.proto, line 37 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66013#-MbG0Hi2-JOHiUAiIbK6:-MbGjsN6DIqNMkILjxWX:b3y7m8f) ([raw file](https://github.com/cockroachdb/cockroach/blob/bbc300dc320aa45099b7f32784d8310cb7b39b29/pkg/sql/execinfrapb/processors_changefeeds.proto#L37)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

[nit] for other `repeated` fields we use plural field names.
</blockquote></details>

Watch_checkpoints is more confusing to me -- I have a single checkpoint.   I blame english.   I introduced Checkpoint message to remove ambiguity.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> Reviewed 6 of 6 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66013)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @miretskiy, and @stevendanna)

___
*[pkg/sql/execinfrapb/processors_changefeeds.proto, line 41 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66013#-MbHSIxT4y1_W3jCz3XM:-MbHSIxUEX9tOOGeHqbP:bs9vd0s) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c531c79000a81b6bd0952ead65a3914fe60dfb7/pkg/sql/execinfrapb/processors_changefeeds.proto#L41)):*
> ```protobuf
> 
>   repeated Watch watches = 1 [(gogoproto.nullable) = false];
>   optional Checkpoint checkpoint  = 5 [(gogoproto.nullable) = false];
> ```

It would be nice to explain the semantics of `checkpoint` here and how its timestamps pertain to the timestamps in `watches`. I worry that the timestamps in checkpoint are more or less meaningless. I also suspect that if we have a checkpoint, then we expect that all of the timestamps inside `watches` to be the exact same value. Should `Checkpoint` just be `repeated roachpb.Span`? The reason to not do it that way would be something forward-looking where we might have different checkpoints for different tables but YAGNI.

Either way, please give this proto a clear explanation.


<!-- Sent from Reviewable.io -->

Commit:
changefeedccl: Add checkpoint field to changefeed progress.

Add `Checkpoint` to ChangefeedProgress.  A checkpoint specifies
the list of spans ahead of the changefeed high water mark.
The checkpointed spans will be skipped during initial backfill
if changefeed restarts.

Release Notes: None

miretskiy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66013)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @stevendanna)

___
*[pkg/sql/execinfrapb/processors_changefeeds.proto, line 41 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66013#-MbHSIxT4y1_W3jCz3XM:-MbIR7m2BGRhLWXy3Q6v:b-t9syd4) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c531c79000a81b6bd0952ead65a3914fe60dfb7/pkg/sql/execinfrapb/processors_changefeeds.proto#L41)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

It would be nice to explain the semantics of `checkpoint` here and how its timestamps pertain to the timestamps in `watches`. I worry that the timestamps in checkpoint are more or less meaningless. I also suspect that if we have a checkpoint, then we expect that all of the timestamps inside `watches` to be the exact same value. Should `Checkpoint` just be `repeated roachpb.Span`? The reason to not do it that way would be something forward-looking where we might have different checkpoints for different tables but YAGNI.

Either way, please give this proto a clear explanation.
</blockquote></details>

You're right.  New version of the code pushed.  Changes made in jobs.proto instead.  I think I maybe able to accomplish all that's needed w/out having to change change aggregator spec.


<!-- Sent from Reviewable.io -->

miretskiy commented:
@ajwerner @stevendanna ready for another look.

ajwerner commented:
It'd be nice to see how this all comes together.

miretskiy commented:
Okay -- i'll just keep this PR open -- probably add more things on top of
this branch as more pieces fall into place.

On Mon, Jun 7, 2021 at 10:54 PM ajwerner ***@***.***> wrote:

> It'd be nice to see how this all comes together.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/66013#issuecomment-856397962>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANA4FVF3U6PFB3ST6F7VQ53TRWA5XANCNFSM457PTJWQ>
> .
>

Commit:
changefeedccl: Use checkpoint when performing initial KV scan.

Make it possible to filter out previously completed spans when
performing initial KV feed scan.

Release Notes: None

Commit:
changefeedccl: Plumb checkpoint argument through distrubted changefeed.

Add "checkpoint" field for change aggregator specification, and plumb
changefeed level checkpoint argument through distributed changefeed
flow.

Implementation using checkpoint in follow on PR.

Release Notes: None

Commit:
changefeedccl: Checkpoint changefeed during backfill.

Checkpoint changefeed backfill progress.  This is accomplished
by resolved span events for each sub-span of the table.
Those events are aggregated by change frontier processor which checkpoints
spans whose timestamp is above the current high water mark.

Release Notes: Changefeed can resume during backfill without
loosing too much progress.

miretskiy commented:
@ajwerner @stevendanna  I think this is ready for a "real" look now.

ajwerner commented on pkg/ccl/changefeedccl/changefeed_test.go:
nit: period.

ajwerner commented on pkg/ccl/changefeedccl/changefeed_test.go:
This feels like another place where the `span.Frontier` is overkill and a `SpanGroup` will be cleaner.

ajwerner commented on pkg/ccl/changefeedccl/changefeed_processors.go:
```diff
@@ -1493,6 +1537,26 @@ func (f *schemaChangeFrontier) SpanFrontier() *span.Frontier {
+	// Span frontier merges adjacent spans, so worst case we have 10K spans.
+	// Each span is a pair of keys.  Those could be large.  Assume 1/2K per key.
+	// So, 1KB per span.  We could be looking at 10MB checkpoint record.
+	// Since we checkpoint infrequently, a 10MB write might be okay.
```
I'd love some better control here. 10MiB scares me. It feels to me like adding some knob on the max size would be very valuable. I'd do it just by throwing away later spans. You'll still make progress so long as you get one resolved span from the worker with the lexicographically highest span because it'll be adjacent. How about a cluster setting and then you just stop appending spans once you hit the threshold?

ajwerner reviewed (commented):
Nice, I like it.

stevendanna commented on pkg/ccl/changefeedccl/changefeed_processors.go:
```diff
@@ -1175,56 +1198,85 @@ func (cf *changeFrontier) noteResolvedSpan(d rowenc.EncDatum) error {
-func (cf *changeFrontier) checkpointResolvedTimestamp(
-	resolved hlc.Timestamp, isBehind bool,
+// checkpointJobProgress checkpoints a changefeed-level job information.
+// In addition, f 'manageProtected' is true, which only happens when frontier advanced,
```
```suggestion
// In addition, if 'manageProtected' is true, which only happens when frontier advanced,
```

miretskiy replied:
Yes; 10MB is scary.  Question is: another PR to specifically target this deficiency vs this pr...

stevendanna commented on pkg/ccl/changefeedccl/changefeed_processors.go:
```diff
@@ -1492,6 +1537,26 @@ func (f *schemaChangeFrontier) SpanFrontier() *span.Frontier {
 	return f.spanFrontier.Frontier
 }
 
+func (f *schemaChangeFrontier) checkpoint(c *jobspb.ChangefeedProgress_Checkpoint) {
```
nit: naming: Might be nice to name this something getCheckpoint or populateCheckpoint to make it clear that c is an out param from the call site.

stevendanna commented on pkg/ccl/changefeedccl/kvfeed/kv_feed.go:
```diff
@@ -296,6 +338,16 @@ func (f *kvFeed) scanIfShould(
+	if initialScan {
+		filtered, err := filterCheckpointSpans(spansToBackfill, f.checkpoint)
+		if err != nil {
+			return err
```
We don't wrap errors in this function, but perhaps we should start?

stevendanna commented on pkg/ccl/changefeedccl/kvfeed/kv_feed.go:
```diff
@@ -185,6 +190,8 @@ func (f *kvFeed) run(ctx context.Context) (err error) {
 			return err
 		}
+
+		// TODO(yevgeniy): Consider using checkpoint here as well.
```
Is the thought of this TODO that we could re-populate the checkpoint information in the case of a schema backfill?

stevendanna reviewed (commented):
Fun to see this come together. I left some minor comments but didn't find anything blocking.  I'll try to give it another read-through.

I like Andrew's idea of putting a configurable cap on the size of the checkpoint.

miretskiy reviewed (commented):
> Reviewed 2 of 9 files at r5, 16 of 17 files at r6, 1 of 1 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66013)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @stevendanna)

___
*[pkg/ccl/changefeedccl/changefeed_processors.go, line 1264 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/66013#-Mbm2EVsGBdY7Cy9q-pr-r7-1264:-McPU3JxCbooldHlrn_n:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cad663aa4c1222b119922703d1ee68ecc2232bd0/pkg/ccl/changefeedccl/changefeed_processors.go#L1264)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

```suggestion
// In addition, if 'manageProtected' is true, which only happens when frontier advanced,
```
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/changefeed_processors.go, line 1546 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/66013#-Mbm2EVsGBdY7Cy9q-pr-r7-1546:-McPUH78-eunGO2_XH7f:b1nl5ge) ([raw file](https://github.com/cockroachdb/cockroach/blob/cad663aa4c1222b119922703d1ee68ecc2232bd0/pkg/ccl/changefeedccl/changefeed_processors.go#L1546)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I'd love some better control here. 10MiB scares me. It feels to me like adding some knob on the max size would be very valuable. I'd do it just by throwing away later spans. You'll still make progress so long as you get one resolved span from the worker with the lexicographically highest span because it'll be adjacent. How about a cluster setting and then you just stop appending spans once you hit the threshold? 
</blockquote></details>

Yeah.  I agree.  I was thinking about that myself.
Basically, if we assume 25 hr GC interval, and a backfill of significantly large table, which takes say 8 hours, we'd be looking at 6*8*10M key...  Absolutely horrible.

So, I think we have to set a target size and work our way back.  Say, we're okay w/ that job record being 150MB across all versions (over 25 hrs).  Then, if we assume worst case (25 hours), then we'd be looking at max of 1 MB per checkpoint (6 MB/hour, 6 checkpoints per hour).  

Another (complimentary idea) is to store checkpoint as bytes -- I'm fairly certain that we loose a lot by not compressing the checkpoint itself -- even though the spans are non contiguous, they do start with the same prefix.

@ajwerner do you think this can go in as a follow on PR, or does it need to be done as part of this pr?
If follow on work focused on making those checkpoints smaller is okay with you, I can reword TODO comment
to be stronger (i.e. this must be done).  WDYT?
___
*[pkg/ccl/changefeedccl/changefeed_test.go, line 4075 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/66013#-Mc1KraUrJ3vH_gFQzQF-r7-4075:-McPbfBh3SOLnLWUUED4:b-blqct4) ([raw file](https://github.com/cockroachdb/cockroach/blob/cad663aa4c1222b119922703d1ee68ecc2232bd0/pkg/ccl/changefeedccl/changefeed_test.go#L4075)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

This feels like another place where the `span.Frontier` is overkill and a `SpanGroup` will be cleaner. 
</blockquote></details>

I wasn't aware of that data structure.  Thanks & done.
___
*[pkg/ccl/changefeedccl/testing_knobs.go, line 32 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/66013#-Mc1KraUrJ3vH_gFQzQH-r7-32:-McPW0NvDTs0GEE4xpuJ:bd49sie) ([raw file](https://github.com/cockroachdb/cockroach/blob/cad663aa4c1222b119922703d1ee68ecc2232bd0/pkg/ccl/changefeedccl/testing_knobs.go#L32)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

`s/shoujld/should/`


I find this knob weird because if it's not set, then we *do* emit. What if instead this was called "shouldSkipResolved" that returns true if the resolved should be skipped. Then, it's more obvious that if the knob doesn't exist then the resolved span will be emitted. 

</blockquote></details>

Ack. Done.
___
*[pkg/ccl/changefeedccl/kvfeed/kv_feed.go, line 346 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/66013#-Mbgx0lCMUwFNdn2JHds-r7-346:-McPXFcZ3PqIFB_wZSLB:ba42lxp) ([raw file](https://github.com/cockroachdb/cockroach/blob/cad663aa4c1222b119922703d1ee68ecc2232bd0/pkg/ccl/changefeedccl/kvfeed/kv_feed.go#L346)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

We don't wrap errors in this function, but perhaps we should start?
</blockquote></details>

Probably better to do it in another PR if you believe that we should be doing this consistently?
___
*[pkg/ccl/changefeedccl/kvfeed/scanner.go, line 151 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66013#-MbG1lct-FwEdXTbXu5d:-Mc4fSpV1_cljtl89P7J:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2f8b03c9328d6e134ddd1478c6c512d7699c40c/pkg/ccl/changefeedccl/kvfeed/scanner.go#L151)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Yup -- good point
</blockquote></details>

Done.
___
*[pkg/sql/execinfrapb/processors_changefeeds.proto, line 37 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66013#-MbG0Hi2-JOHiUAiIbK6:-Mc4fFag0dwYUfYGZ8yE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bbc300dc320aa45099b7f32784d8310cb7b39b29/pkg/sql/execinfrapb/processors_changefeeds.proto#L37)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Watch_checkpoints is more confusing to me -- I have a single checkpoint.   I blame english.   I introduced Checkpoint message to remove ambiguity.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

miretskiy commented on pkg/ccl/changefeedccl/kvfeed/scanner.go:
```diff
@@ -120,22 +125,28 @@ func (p *scanRequestScanner) exportSpan(
+		remaining = res.ResumeSpan
 	}
 	// p.metrics.PollRequestNanosHist.RecordValue(scanDuration.Nanoseconds())
 	if err := sink.AddResolved(ctx, span, ts, jobspb.ResolvedSpan_NONE); err != nil {
```
Yeah, we still need it -- we might not hit the AddResolved above (i..e span is small enough to be exported in 1 shot).
I could emit the resolved just for the last part that wasn't done, but that seems like an overkill.

miretskiy commented on pkg/ccl/changefeedccl/kvfeed/kv_feed.go:
```diff
@@ -243,6 +250,35 @@ func isRegionalByRowChange(events []schemafeed.TableEvent) bool {
+			return span.ContinueMatch
+		})
+	}
+	return filtered, nil
```
I think it's a good idea.  Question though: Add a TODO and send another PR adding Sub method and switching to that; vs doing it as part of these PRS?
My preferences is for the todo.

miretskiy commented:
@ajwerner @stevendanna  Rebased; addressed all of your comments -- few questions re TODOs and fixes in follow on PRs.

Commit:
roachpb: Add `Sub` method to `SpanGroup`

Add `Sub` method which removes set of spans from the `SpanGroup.

Release Notes: None

miretskiy commented:
@ajwerner @stevendanna FYI: Addressed remaining todos.
  * Added commit to this PR  to add `Sub` to `SpanGroup`.
  * Use `SpanGroup` when computing spans to backfill.
  * Added limit setting for maximum checkpoint size
  * Augmented checkpoint test to also test w/ limit.

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 4 of 9 files at r5, 2 of 13 files at r8, 2 of 11 files at r9, 3 of 8 files at r11, 11 of 12 files at r12, 1 of 1 files at r13.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66013#-:-MciiGhKBEgmwqFxcjUM:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @miretskiy, and @stevendanna)

___
*[pkg/ccl/changefeedccl/changefeed_processors.go, line 1546 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/66013#-McigwTW-siVrdGjc5fJ:-McigwTW-siVrdGjc5fK:b-o5wx0m) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd605f1d85cc7561b04026ecc867b2be68f78bf8/pkg/ccl/changefeedccl/changefeed_processors.go#L1546)):*
> ```
> 
> 	maxBytes int64, c *jobspb.ChangefeedProgress_Checkpoint,
> ) {
> ```
This deserves commentary indicating that the checkpoint is being modified in place. It's not really expected in go the way this would be in C++. TBH I don't see why you'd want to do this like this anyway. It's weird that the caller needs to zero out the checkpoint before passing it in. Why not just return `jobspb.ChangefeedProgress_Checkpoint`? Is the idea that you want to re-use the backing slice? That seems like premature optimization given all the allocations involved in changefeeds. 


<!-- Sent from Reviewable.io -->

miretskiy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66013)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner and @stevendanna)

___
*[pkg/ccl/changefeedccl/changefeed_processors.go, line 1546 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/66013#-McigwTW-siVrdGjc5fJ:-Mcil3dp9jA78nB0GX6p:b-iboxvo) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd605f1d85cc7561b04026ecc867b2be68f78bf8/pkg/ccl/changefeedccl/changefeed_processors.go#L1546)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 
> 	maxBytes int64, c *jobspb.ChangefeedProgress_Checkpoint,
> ) {
> ```
This deserves commentary indicating that the checkpoint is being modified in place. It's not really expected in go the way this would be in C++. TBH I don't see why you'd want to do this like this anyway. It's weird that the caller needs to zero out the checkpoint before passing it in. Why not just return `jobspb.ChangefeedProgress_Checkpoint`? Is the idea that you want to re-use the backing slice? That seems like premature optimization given all the allocations involved in changefeeds. 
</blockquote></details>

No idea what I was thinking. Switched to return `[]roachpb.Span`


<!-- Sent from Reviewable.io -->

miretskiy commented:
tftr
bors r+
